### PR TITLE
Track worker version & auto-disable

### PR DESCRIPTION
**Pull request by peterebden**
This is about the lightest weight possible option for a changing fleet; we set a min proportion that a particular version has to reach of the fleet before it becomes enabled. Hence it's adaptive to fleet size and allows an incoming set to become active once they exceed some proportion.

During rollouts there would be a period of reduced capacity before the incoming version picked up enough mass to start to take over, but that's unlikely to be a big issue if the proportion is set relatively low (e.g. 10 or 20% of capacity), assuming it's not happening at a peak time.

### CODE REVIEW

**Review by geonaut — COMMENTED** [review]
A little advanced for me, but think it makes sense. Nothing major that I spotted

File: lucidity/BUILD [hunk 1]
```diff
@@ -12,5 +12,5 @@ go_binary(
 sh_cmd(
     name = "run_local",
     srcs = [":lucidity"],
-    cmd = "exec $(out_location :lucidity) --host 127.0.0.1 --port 7774 --admin_disabled",
+    cmd = "exec $(out_location :lucidity) --host 127.0.0.1 --port 7774 --http_port 7773 --admin_disabled -v info --min_proportion 0.5",
```
[the remark this answers is not in the document]
Comment (peterebden) [reply lucidity/BUILD:15]: Not really, I just keep forgetting what port it is and looking at this file, then going to 7774 (which is the gRPC port, so it fails).

File: lucidity/rpc/rpc.go [hunk 2]
```diff
@@ -184,3 +190,34 @@ func (s *server) removeWorker(key string) {
 	delete(s.unhealthyWorkers, key)
 	delete(s.busyWorkers, key)
 }
+
+// recalculateValidVersions runs through all the workers and calculates the new set of versions we'll allow to be active.
```
[the remark this answers is not in the document]
Comment (peterebden) [reply lucidity/rpc/rpc.go:201]: Good point, and mostly just laziness. Have added one.

File: lucidity/rpc/rpc.go [hunk 3]
```diff
@@ -184,3 +190,34 @@ func (s *server) removeWorker(key string) {
...
+		counts[v.(*pb.UpdateRequest).Version]++
+		n++
+		return true
+	})
+	// Now add any of those over the threshold to the set of valid versions
+	min := int(s.minProportion * float64(n))
+	log.Notice("Updating valid versions, total workers: %d, min threshold: %d", n, min)
+	for k, v := range counts {
+		if v < min {
+			log.Info("  %s: %d (invalid)", k, v)
```
[the remark this answers is not in the document]
Comment (peterebden) [reply lucidity/rpc/rpc.go:216]: Yeah it's https://pkg.go.dev/github.com/op/go-logging (which we used to use internally too before we moved to Logrus; this one is fairly dead but continues to work fine)

**Pull request merged**